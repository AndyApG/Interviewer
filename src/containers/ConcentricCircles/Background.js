import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
import Radar from './Radar';
import { makeGetSociogramOptions } from '../../selectors/sociogram';
import { BackgroundImage } from '../../components';
import sociogramOptionsProps from './propTypes';

class Background extends PureComponent {
  static propTypes = {
    ...sociogramOptionsProps,
  };

  static defaultProps = {
    concentricCircles: 4,
    skewedTowardCenter: true,
    image: null,
  };

  render() {
    const { concentricCircles, skewedTowardCenter, image } = this.props;
    let background;

    if (image) {
      background = <BackgroundImage className="sociogram-background__image" url={image} />;
    } else {
      background = <Radar n={concentricCircles} skewed={skewedTowardCenter} />;
    }

    return (
      <div className="sociogram-background">
        { background }
      </div>
    );
  }
}

function makeMapStateToProps() {
  const getSociogramOptions = makeGetSociogramOptions();

  return function mapStateToProps(state, props) {
    return {
      ...getSociogramOptions(state, props),
    };
  };
}

export { Background };

export default connect(makeMapStateToProps)(Background);
