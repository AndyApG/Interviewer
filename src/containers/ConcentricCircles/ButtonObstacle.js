import React, { PureComponent } from 'react';
import { withHandlers, compose } from 'recompose';
import PropTypes from 'prop-types';

import { Button } from '../../ui/components';
import { withBounds } from '../../behaviours';
import { DropObstacle } from '../../behaviours/DragAndDrop';

class ButtonObstacle extends PureComponent {
  render() {
    const {
      label,
      accepts,
      ...rest
    } = this.props;

    return (
      <Button
        {...rest}
      >
        {label}
      </Button>
    );
  }
}

ButtonObstacle.propTypes = {
  accepts: PropTypes.func.isRequired,
  label: PropTypes.string,
};

ButtonObstacle.defaultProps = {
  label: '',
};

export default compose(
  withBounds,
  withHandlers({
    accepts: () => () => true,
  }),
  DropObstacle,
)(ButtonObstacle);
