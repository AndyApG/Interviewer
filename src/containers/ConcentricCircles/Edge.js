import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { makeGetEdgeColor } from '../../selectors/protocol';

const getEdgeColor = makeGetEdgeColor();

export class Edge extends PureComponent {
  render() {
    const { from, to, color } = this.props;
    if (!from || !to) { return null; }

    return (
      <line
        x1={from.x}
        y1={from.y}
        x2={to.x}
        y2={to.y}
        stroke={`var(--${color})`}
      />
    );
  }
}

function mapStateToProps(state, props) {
  return {
    color: getEdgeColor(state, props),
  };
}

Edge.propTypes = {
  color: PropTypes.string,
  from: PropTypes.object.isRequired,
  to: PropTypes.object.isRequired,
  /* eslint-disable react/no-unused-prop-types */
  type: PropTypes.string.isRequired,
};

Edge.defaultProps = {
  color: 'edge-color-seq-1',
};

export default connect(mapStateToProps)(Edge);
