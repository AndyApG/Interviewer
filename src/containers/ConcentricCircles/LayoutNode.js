import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { compose } from 'recompose';
import { Node } from 'network-canvas-ui';
import { selectable } from '../../behaviours';
import { DragSource } from '../../behaviours/DragAndDrop';

const label = node => node.nickname;

const EnhancedNode = compose(
  DragSource,
  selectable,
)(Node);

class LayoutNode extends PureComponent {
  render() {
    const {
      node,
      selected,
      allowPositioning,
      allowSelect,
      layoutVariable,
    } = this.props;

    const { x, y } = node[layoutVariable];

    const styles = {
      left: `${100 * x}%`,
      top: `${100 * y}%`,
      transform: 'translate(-50%, -50%)',
    };

    return (
      <div
        className="node-layout__node"
        style={styles}
      >
        <EnhancedNode
          label={label(node)}
          onSelected={this.props.onSelected}
          selected={selected}
          allowDrag={allowPositioning}
          allowSelect={allowSelect}
          meta={() => ({ ...node, itemType: 'POSITIONED_NODE' })}
          animate={false}
          {...node}
        />
      </div>
    );
  }
}

LayoutNode.propTypes = {
  onSelected: PropTypes.func.isRequired,
  layoutVariable: PropTypes.string.isRequired,
  node: PropTypes.object.isRequired,
  allowPositioning: PropTypes.bool,
  allowSelect: PropTypes.bool,
  selected: PropTypes.bool,
  areaWidth: PropTypes.number,
  areaHeight: PropTypes.number,
};

LayoutNode.defaultProps = {
  allowPositioning: false,
  allowSelect: false,
  selected: false,
  areaWidth: 0,
  areaHeight: 0,
  onSelected: () => {},
};

export { LayoutNode };

export default LayoutNode;
