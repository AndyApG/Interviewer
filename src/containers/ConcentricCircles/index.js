import ConcentricCircles from './ConcentricCircles';

export default ConcentricCircles;
