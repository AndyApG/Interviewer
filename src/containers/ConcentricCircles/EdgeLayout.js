import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Edge } from './Edge';
import { makeDisplayEdgesForPrompt } from '../../selectors/sociogram';

export class EdgeLayout extends PureComponent {
  static propTypes = {
    displayEdges: PropTypes.array,
  };

  static defaultProps = {
    displayEdges: [],
  };

  renderEdge = edge => (<Edge {...edge} />);


  render() {
    const { displayEdges } = this.props;
    return (
      <div className="edge-layout">
        <svg viewBox="0 0 1 1" xmlns="http://www.w3.org/2000/svg" preserveAspectRatio="none">
          { displayEdges.map(this.renderEdge) }
        </svg>
      </div>
    );
  }
}

function makeMapStateToProps() {
  const displayEdgesForPrompt = makeDisplayEdgesForPrompt();

  return function mapStateToProps(state, props) {
    return {
      displayEdges: displayEdgesForPrompt(state, props),
    };
  };
}

export default connect(makeMapStateToProps)(EdgeLayout);
