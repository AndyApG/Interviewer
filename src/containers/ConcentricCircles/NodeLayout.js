import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { compose, withHandlers, withState } from 'recompose';
import { isEqual, isEmpty, pick, isMatch, has } from 'lodash';
import LayoutNode from './LayoutNode';
import { withBounds } from '../../behaviours';
import { makeGetSociogramOptions, makeGetPlacedNodes } from '../../selectors/sociogram';
import { actionCreators as sessionsActions } from '../../ducks/modules/sessions';
import { nodePrimaryKeyProperty, getNodeAttributes } from '../../ducks/modules/network';
import { DropTarget } from '../../behaviours/DragAndDrop';
import sociogramOptionsProps from './propTypes';

const watchProps = ['width', 'height', 'dropCount'];

const propsChangedExcludingNodes = (nextProps, props) =>
  !isEqual(pick(nextProps, watchProps), pick(props, watchProps));

const nodesLengthChanged = (nextProps, props) =>
  nextProps.nodes.length !== props.nodes.length;

const relativeCoords = (container, node) => ({
  x: (node.x - container.x) / container.width,
  y: (node.y - container.y) / container.height,
});

const dropHandlers = compose(
  withState('dropCount', 'setDropCount', 0),
  withHandlers({
    accepts: () => ({ meta }) => meta.itemType === 'POSITIONED_NODE',
    onDrop: props => (item) => {
      props.updateNode(
        {
          [nodePrimaryKeyProperty]: item.meta[nodePrimaryKeyProperty],
        },
        {
          [props.layoutVariable]: relativeCoords(props, item),
        },
      );

      // Horrible hack for performance (only re-render nodes on drop, not on drag)
      props.setDropCount(props.dropCount + 1);
    },
    onDrag: props => (item) => {
      if (!has(item.meta, props.layoutVariable)) { return; }

      props.updateNode(
        {
          [nodePrimaryKeyProperty]: item.meta[nodePrimaryKeyProperty],
        },
        {
          [props.layoutVariable]: relativeCoords(props, item),
        },
      );
    },
  }),
);

class NodeLayout extends Component {
  static propTypes = {
    nodes: PropTypes.array,
    toggleEdge: PropTypes.func.isRequired,
    toggleHighlight: PropTypes.func.isRequired,
    ...sociogramOptionsProps,
  };

  static defaultProps = {
    nodes: [],
    allowPositioning: true,
    allowSelect: true,
  };

  constructor(props) {
    super(props);

    this.state = {
      connectFrom: null,
    };
  }

  shouldComponentUpdate(nextProps) {
    if (nodesLengthChanged(nextProps, this.props)) { return true; }
    if (propsChangedExcludingNodes(nextProps, this.props)) { return true; }

    return false;
  }

  onSelected = (node) => {
    const {
      allowSelect,
    } = this.props;

    if (!allowSelect) { return; }

    this.connectNode(node[nodePrimaryKeyProperty]);

    this.toggleHighlightAttributes(node[nodePrimaryKeyProperty]);

    this.forceUpdate();
  }

  connectNode(nodeId) {
    const { createEdge, canCreateEdge } = this.props;
    const { connectFrom } = this.state;

    if (!canCreateEdge) { return; }

    if (!connectFrom) {
      this.setState({ connectFrom: nodeId });
      return;
    }

    if (connectFrom !== nodeId) {
      this.props.toggleEdge({
        from: connectFrom,
        to: nodeId,
        type: createEdge,
      });
    }

    this.setState({ connectFrom: null });
  }

  toggleHighlightAttributes(nodeId) {
    if (!this.props.allowHighlighting) { return; }

    this.props.toggleHighlight(
      nodeId,
      { ...this.props.highlightAttributes },
    );
  }

  isHighlighted(node) {
    return !isEmpty(this.props.highlightAttributes) &&
      isMatch(node, this.props.highlightAttributes);
  }

  isLinking(node) {
    return this.props.allowSelect &&
      this.props.canCreateEdge &&
      node[nodePrimaryKeyProperty] === this.state.connectFrom;
  }

  render() {
    const {
      nodes,
      allowPositioning,
      allowSelect,
      layoutVariable,
    } = this.props;

    return (
      <div className="node-layout">
        { nodes.map((node) => {
          const nodeAttributes = getNodeAttributes(node);
          if (!has(nodeAttributes, layoutVariable)) { return null; }

          return (
            <LayoutNode
              key={node[nodePrimaryKeyProperty]}
              node={node}
              layoutVariable={layoutVariable}
              onSelected={() => this.onSelected(node)}
              selected={this.isHighlighted(node)}
              linking={this.isLinking(node)}
              allowPositioning={allowPositioning}
              allowSelect={allowSelect}
              areaWidth={this.props.width}
              areaHeight={this.props.height}
            />
          );
        }) }
      </div>
    );
  }
}

function makeMapStateToProps() {
  const getPlacedNodes = makeGetPlacedNodes();
  const getSociogramOptions = makeGetSociogramOptions();

  return function mapStateToProps(state, props) {
    return {
      nodes: getPlacedNodes(state, props),
      ...getSociogramOptions(state, props),
    };
  };
}

function mapDispatchToProps(dispatch) {
  return {
    toggleHighlight: bindActionCreators(sessionsActions.toggleNodeAttributes, dispatch),
    toggleEdge: bindActionCreators(sessionsActions.toggleEdge, dispatch),
    updateNode: bindActionCreators(sessionsActions.updateNode, dispatch),
  };
}

export { NodeLayout };

export default compose(
  connect(makeMapStateToProps, mapDispatchToProps),
  withBounds,
  dropHandlers,
  DropTarget,
)(NodeLayout);
