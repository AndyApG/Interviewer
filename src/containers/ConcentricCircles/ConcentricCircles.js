import React from 'react';
import PropTypes from 'prop-types';
import { NodeBucket } from '../../containers';
import NodeLayout from './NodeLayout';
import EdgeLayout from './EdgeLayout';
import Background from './Background';

const Sociogram = ({ stage, prompt }) => (
  <div className="sociogram">
    <Background stage={stage} prompt={prompt} />
    {
      prompt.edges &&
      <EdgeLayout
        stage={stage}
        prompt={prompt}
      />
    }
    <NodeLayout
      id="NODE_LAYOUT"
      stage={stage}
      prompt={prompt}
    />
    <NodeBucket
      stage={stage}
      prompt={prompt}
    />
  </div>
);

Sociogram.propTypes = {
  stage: PropTypes.object.isRequired,
  prompt: PropTypes.object.isRequired,
};

export { Sociogram };

export default Sociogram;
