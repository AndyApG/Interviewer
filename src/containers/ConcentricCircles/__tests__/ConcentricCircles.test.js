/* eslint-env jest */

import React from 'react';
import { shallow } from 'enzyme';
import ConcentricCircles from '../ConcentricCircles';

const mockProps = {
  prompt: { sociogram: { background: {} } },
  stage: {},
};

describe('<ConcentricCircles />', () => {
  it('renders ok', () => {
    const component = shallow(<ConcentricCircles {...mockProps} />);

    expect(component).toMatchSnapshot();
  });
});
