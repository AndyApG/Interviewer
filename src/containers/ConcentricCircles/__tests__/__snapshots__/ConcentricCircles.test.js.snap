// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ConcentricCircles /> renders ok 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Sociogram
    prompt={
        Object {
            "sociogram": Object {
              "background": Object {},
            },
          }
    }
    stage={Object {}}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Connect(Background)
          prompt={
                    Object {
                              "sociogram": Object {
                                "background": Object {},
                              },
                            }
          }
          stage={Object {}}
/>,
        undefined,
        <Connect(Component)
          id="NODE_LAYOUT"
          prompt={
                    Object {
                              "sociogram": Object {
                                "background": Object {},
                              },
                            }
          }
          stage={Object {}}
/>,
        <Connect(NodeBucket)
          prompt={
                    Object {
                              "sociogram": Object {
                                "background": Object {},
                              },
                            }
          }
          stage={Object {}}
/>,
      ],
      "className": "sociogram",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "prompt": Object {
            "sociogram": Object {
              "background": Object {},
            },
          },
          "stage": Object {},
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      undefined,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "id": "NODE_LAYOUT",
          "prompt": Object {
            "sociogram": Object {
              "background": Object {},
            },
          },
          "stage": Object {},
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "prompt": Object {
            "sociogram": Object {
              "background": Object {},
            },
          },
          "stage": Object {},
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Connect(Background)
            prompt={
                        Object {
                                    "sociogram": Object {
                                      "background": Object {},
                                    },
                                  }
            }
            stage={Object {}}
/>,
          undefined,
          <Connect(Component)
            id="NODE_LAYOUT"
            prompt={
                        Object {
                                    "sociogram": Object {
                                      "background": Object {},
                                    },
                                  }
            }
            stage={Object {}}
/>,
          <Connect(NodeBucket)
            prompt={
                        Object {
                                    "sociogram": Object {
                                      "background": Object {},
                                    },
                                  }
            }
            stage={Object {}}
/>,
        ],
        "className": "sociogram",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "prompt": Object {
              "sociogram": Object {
                "background": Object {},
              },
            },
            "stage": Object {},
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        undefined,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "id": "NODE_LAYOUT",
            "prompt": Object {
              "sociogram": Object {
                "background": Object {},
              },
            },
            "stage": Object {},
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "prompt": Object {
              "sociogram": Object {
                "background": Object {},
              },
            },
            "stage": Object {},
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
